### PR TITLE
#24 - add Link and BinBin providers

### LINKED ISSUE #24: [Providers] Add Link provider
The aim of this task is to import Link provider to the application.
Link to provider's site: https://superpedestrian.com/ (I think so)

### LINKED ISSUE #92: Add BinBin provider
Site: https://www.binbin.tech/lokasyonlar

### PR DESCRIPTION
**Pull request by AleksandraKozubal**
Added translation method in DataImporter, I think it could be helpful to solve issue #104.

It should close #24, and close #92

### CODE REVIEW

File: app/Importers/DataImporter.php [hunk 1]
```diff
@@ -48,6 +49,13 @@ public static function getProviderName(): string
         return $classNameParts[0];
     }
 
+    public function translate(string $word, $language): string
+    {
+        $translate = new GoogleTranslate($language);
```
Comment (krzysztofrewak) [inline app/Importers/DataImporter.php:54]: Could you inject it to the constructor?

File: app/Importers/BinBinDataImporter.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                $cityName = $this->translate(trim($data[0]), "en");
+            }
+
+            $provider = $this->load($cityName, $countryName);
+
+            if ($provider !== "") {
+                $existingCityProviders[] = $provider;
+            }
+        }
+        $this->deleteMissingProviders(self::getProviderName(), $existingCityProviders);
```
Comment (krzysztofrewak) [inline app/Importers/BinBinDataImporter.php:70]:
```suggestion
        }

        $this->deleteMissingProviders(self::getProviderName(), $existingCityProviders);
```

File: app/Importers/BinBinDataImporter.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+        $existingCityProviders = [];
+
+        foreach ($this->sections as $section) {
+            $data = explode("|", $section->nodeValue);
+
+            if (trim($data[1]) === "Türkiye") {
+                $countryName = "Turkey";
+            } else {
+                $countryName = $this->translate(trim($data[1]), "en");
```
Comment (krzysztofrewak) [inline app/Importers/BinBinDataImporter.php:53]: Please avoid using strings like `"en"` directly in your code. Maybe assiging it to some constant would be better approach?

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

**Comment by krzysztofrewak** [comment]
Also: listing available providers in readme file would be a nice thing.

File: app/Importers/DataImporter.php (hunk 1 above)
Comment (AleksandraKozubal) [reply app/Importers/DataImporter.php]: sure, done

File: app/Importers/BinBinDataImporter.php (hunk 2 above)
Comment (AleksandraKozubal) [reply app/Importers/BinBinDataImporter.php]: added

File: app/Importers/BinBinDataImporter.php (hunk 3 above)
Comment (AleksandraKozubal) [reply app/Importers/BinBinDataImporter.php]: moved to constant

File: app/Importers/DataImporter.php (hunk 1 above)
Comment (AleksandraKozubal) [reply app/Importers/DataImporter.php]: It creates a problem with ImporterTest - GoogleTranslate is a final class, it can't be mocked

Comment (AleksandraKozubal) [reply app/Importers/DataImporter.php]: done

File: app/Importers/BinBinDataImporter.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        $existingCityProviders = [];
+
+        foreach ($this->sections as $section) {
+            $data = explode("|", $section->nodeValue);
+
+            if (trim($data[1]) === "Türkiye") $countryName = "Turkey"; else $countryName = $this->translate(trim($data[1]), self::language);
+
+            if (trim($data[0]) === "Uşak") $cityName = "Usak"; elseif (trim($data[0]) === "Murter") continue;
+            else $cityName = $this->translate(trim($data[0]), "en");
```
Comment (kamilpiech97) [inline app/Importers/BinBinDataImporter.php:55]:
```suggestion
            if (trim($data[1]) === "Türkiye") 
                  $countryName = "Turkey"; 
            else 
                  $countryName = $this->translate(trim($data[1]), self::language);

            if (trim($data[0]) === "Uşak") 
                  $cityName = "Usak"; 
            elseif (trim($data[0]) === "Murter") 
                  continue;
            else 
                  $cityName = $this->translate(trim($data[0]), "en");
```

Comment (kamilpiech97) [reply app/Importers/BinBinDataImporter.php:55]: more readable?

File: readme.md [hunk 5]
```diff
@@ -1,4 +1,23 @@
...
+7. Lime
+8. Link
+9. Neuron
+10. Quick
+11. Ryde
+12. Spin
+13. Tier
+13. Urent
+14. Voi
+15. Zwings
```
Comment (krzysztofrewak) [inline readme.md:19]:
```suggestion
1. BinBin
1. Bird  
1. BitMobility
1. Bolt
1. Dott
1. Hulaj
1. Lime
1. Link
1. Neuron
1. Quick
1. Ryde
1. Spin
1. Tier
1. Urent
1. Voi
1. Zwings
```
Markdown supports numbered lists: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#lists

File: app/Importers/BinBinDataImporter.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+            if (trim($data[0]) === "Uşak") 
+            $cityName = "Usak"; 
+            elseif (trim($data[0]) === "Murter") 
+            continue;
+            else $cityName = $this->translate(trim($data[0]), "en");
+
+            $provider = $this->load($cityName, $countryName);
+
+            if ($provider !== "") $existingCityProviders[] = $provider;
```
Comment (krzysztofrewak) [inline app/Importers/BinBinDataImporter.php:64]: This code formatting is fucked up. PHPStorm Ctrl+Alt+L should fix it, but you can also wait until this is merged: https://github.com/blumilksoftware/codestyle/pull/99

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: app/Importers/BinBinDataImporter.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+declare(strict_types=1);
+
+namespace App\Importers;
+
+use GuzzleHttp\Exception\GuzzleException;
+use Symfony\Component\DomCrawler\Crawler;
+
+class BinBinDataImporter extends DataImporter
+{
+    public const language = "en";
```
[the remark this answers is not in the document]
Comment (AleksandraKozubal) [reply app/Importers/BinBinDataImporter.php]: changed

File: app/Importers/BinBinDataImporter.php (hunk 6 above)
Comment (AleksandraKozubal) [reply app/Importers/BinBinDataImporter.php]: fixed

**Review by krzysztofrewak — APPROVED** [review]

File: app/Importers/DataImporter.php [hunk 8]
```diff
@@ -48,6 +52,11 @@ public static function getProviderName(): string
         return $classNameParts[0];
     }
 
+    public function translate(string $word, $language): string
```
Comment (kamilpiech97) [inline app/Importers/DataImporter.php:55]:
```suggestion
    public function translate(string $word, string $language): string
```

File: readme.md [hunk 9]
```diff
@@ -1,4 +1,26 @@
...
+1. Link
+1. Neuron
+1. Quick
+1. Ryde
+1. Spin
+1. Tier
+1. Urent
+1. Veo
+1. Voi
+1. Zwings
```
Comment (kamilpiech97) [inline readme.md:22]:
```suggestion
1. Beam
2. Beryl
3. BinBin
4. Bird
5. BitMobility
6. Bolt
7. Dott
8. Hulaj
9. Lime
10. Link
11. Neuron
12. Quick
13. Ryde
14. Spin
15. Tier
16. Urent
17. Veo
18. Voi
19. Zwings
```

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: readme.md (hunk 9 above)
Comment (AleksandraKozubal) [reply readme.md:22]: This was suggested in https://github.com/blumilksoftware/escooters/pull/117#discussion_r1299739789

File: app/Importers/DataImporter.php (hunk 8 above)
Comment (AleksandraKozubal) [reply app/Importers/DataImporter.php]: changed

**Review by kamilpiech97 — APPROVED** [review]

**Review by Letha0 — APPROVED** [review]

**Pull request merged**